### PR TITLE
add the ``nodecheck`` crash command

### PR DESCRIPTION
**Pull request by kovrus**

### CODE REVIEW

File: src/crate/crash/commands.py [hunk 1]
```diff
@@ -91,47 +91,82 @@ def __call__(self, cmd, *args, **kwargs):
...
+    DEFAULT_STMT = """
+        SELECT n.name AS "Node Name", c.description AS "Failed Checks"
+        FROM sys.node_checks c, sys.nodes n
+        WHERE c.passed = false
+          AND c.acknowledged = false
+          AND c.node_id = n.id
+        ORDER BY c.id, severity asc"""
+
+    STARTUP_STMT = """
+        SELECT description as "Failed Checks", count(*) as "Num. Nodes"
```
Comment (kovrus) [inline src/crate/crash/commands.py]: ``Num. nodes``... we can think about some other column name

File: CHANGES.txt [hunk 2]
```diff
@@ -5,6 +5,9 @@ Changes for crash
 Unreleased
 ==========
 
+ - Added the ``nodecheck`` command which queries the ``sys.node_checks`` table
+   for failing node checks.
```
Comment (chaudum) [inline CHANGES.txt]: Failing node checks are also displayed in a condensed form at startup.

File: src/crate/crash/output.txt [hunk 3]
```diff
@@ -10,7 +10,7 @@ Check the correct response status lines.
     +------------------------+-----------+---------+-----------+---------+
     | http://127.0.0.1:44209 | crate     | ...     | TRUE      | OK      |
     +------------------------+-----------+---------+-----------+---------+
-    CONNECT OKCLUSTER CHECK OK
+    CONNECT OKCLUSTER CHECK OKCrate ... does not support the "nodecheck" command.
```
Comment (chaudum) [inline src/crate/crash/output.txt]: hm, is this expected output? no space/newline?

File: src/crate/crash/commands.py [hunk 4]
```diff
@@ -91,47 +91,82 @@ def __call__(self, cmd, *args, **kwargs):
...
+        success = cmd._execute(stmt)
         cmd.exit_code = cmd.exit_code or int(not success)
         if not success:
             return False
         cur = cmd.cursor
         print_vars = {
             's': 'S'[cur.rowcount == 1:],
-            'rowcount': cur.rowcount
+            'rowcount': cur.rowcount,
+            'msg': msg
```
Comment (chaudum) [inline src/crate/crash/commands.py]: I wouldn't pass the `msg` variable but define a class variable 
and use `self.CHECK_NAME` here

**Comment by kovrus** [comment]
image
image

File: src/crate/crash/commands.py [hunk 5]
```diff
@@ -91,47 +91,89 @@ def __call__(self, cmd, *args, **kwargs):
...
-            cmd.logger.info('CLUSTER CHECK OK')
+            cmd.logger.info('{} OK'.format(self.CHECK_NAME))
         return True
 
 
+class NodeCheckCommand(CheckCommand):
+    """ print failed node checks """
+
+    DEFAULT_STMT = """
+        SELECT n.name AS "Node Name", c.description AS "Failed Checks"
```
Comment (chaudum) [inline src/crate/crash/commands.py]: ```AS "Failed Check"```

File: src/crate/crash/commands.py (hunk 1 above)
Comment (chaudum) [inline src/crate/crash/commands.py]: `AS "Failed Check"`

File: CHANGES.txt [hunk 6]
```diff
@@ -5,6 +5,14 @@ Changes for crash
 Unreleased
 ==========
 
+ - Make the ``\check`` command to display both, node and cluster checks.
+   Allow ``node`` and ``cluster`` arguments for the command to display
+   node and cluster checks, respectively.
+
+ - Added the ``nodecheck`` command which queries the ``sys.node_checks`` table
```
Comment (chaudum) [inline CHANGES.txt]: remove 2nd entry

File: src/crate/crash/commands.py [hunk 7]
```diff
@@ -91,47 +92,105 @@ def __call__(self, cmd, *args, **kwargs):
...
+                ('node', NodeCheckCommand())
+            ])
+
+    def __call__(self, cmd, *args, **kwargs):
+        if not args:
+            [check(cmd, args, kwargs) for check in self.CHECKS.values()]
+        elif len(args) == 1 and args[0] in self.CHECKS:
+            self.CHECKS[args[0]](cmd, *args, **kwargs)
+        else:
+            cmd.logger.warn('Wrong argument for "\check": {}'.format(*args))
```
Comment (chaudum) [inline src/crate/crash/commands.py]: suggestion: No check for {}.

File: src/crate/crash/commands.py [hunk 8]
```diff
@@ -91,47 +92,105 @@ def __call__(self, cmd, *args, **kwargs):
...
+
+
+class CheckCommand(Command):
+    """ print failed cluster and/or node checks, e.g. \check node """
+
+    CHECKS = OrderedDict([
+                ('cluster', ClusterCheckCommand()),
+                ('node', NodeCheckCommand())
+            ])
+
```
Comment (chaudum) [inline src/crate/crash/commands.py]: please also implement the complete() method

File: src/crate/crash/commands.py [hunk 9]
```diff
@@ -91,47 +92,105 @@ def __call__(self, cmd, *args, **kwargs):
...
+    def __call__(self, cmd, *args, **kwargs):
+        if not args:
+            [check(cmd, args, kwargs) for check in self.CHECKS.values()]
+        elif len(args) == 1 and args[0] in self.CHECKS:
+            self.CHECKS[args[0]](cmd, *args, **kwargs)
+        else:
+            cmd.logger.warn('Wrong argument for "\check": {}'.format(*args))
+
+    def exec_check(self, cmd, checks, *args, **kwargs):
```
Comment (chaudum) [inline src/crate/crash/commands.py]: unused

File: src/crate/crash/commands.py [hunk 10]
```diff
@@ -91,47 +92,105 @@ def __call__(self, cmd, *args, **kwargs):
...
+
+class CheckCommand(Command):
+    """ print failed cluster and/or node checks, e.g. \check node """
+
+    CHECKS = OrderedDict([
+                ('cluster', ClusterCheckCommand()),
+                ('node', NodeCheckCommand())
+            ])
+
+    def __call__(self, cmd, *args, **kwargs):
```
Comment (chaudum) [inline src/crate/crash/commands.py]:
```py
def __call__(self, cmd, check_name=None, **kwargs):
    ...
```
like here: https://github.com/crate/crash/pull/142/files#diff-60079d3f543d7fedc2bc3125476434ceR77

**Comment by chaudum** [comment]
lgtm. @bogensberger  what do you think?

**Comment by bogensberger** [comment]
muy bien

**Pull request merged**